ompi/mpi/c/address.c

```diff
@@ -21,6 +21,17 @@
 #include "ompi_config.h"
 #include <stdio.h>
 
+/* This implementation has been removed from the MPI 3.1 standard.
+ * Open MPI v4.0.x is keeping the implementation in the library, but
+ * removing the prototypes from the headers, unless the user configures
+ * with --enable-mpi1-compatibility.
+ *
+ * To prevent having to port these implementations of removed functions
+ * to the newer MPI calls, we are defining OMPI_OMIT_MPI1_COMPAT_DECLS
+ * to 0 before including the c bindings.
+ */
+#define OMPI_OMIT_MPI1_COMPAT_DECLS 0
```

Comment:
Shouldn't the `OMPI_BUILDING` `#define` be sufficient here? If not, why?

Reply:
So, there's currently no logic to affect `OMPI_OMIT_MPI1_COMPAT_DECLS` via `OMPI_BUILDING`.  Are you proposing we add that?

Reply:
`OMPI_BUILDING` already overrules `OMPI_WANT_MPI_INTERFACE_WARNING`, where the latter is for users. Why shouldn't `OMPI_BUILDING` also overrule `OMPI_OMIT_MPI1_COMPAT_DECLS`, and leave `OMPI_OMIT_MPI1_COMPAT_DECLS` to the users?

Reply:
What do you mean by _"leave `OMPI_OMIT_MPI1_COMPAT_DECLS` to the users?"_  Users can control that already via the --enable-mpi-compatibility configure flag.  This is just silently unsetting during the build of a few files in the .c files (not mpi.h) to allow those files to build.

Reply:
As long as the symbols are in the library, I think its valid to access them by there prototype. I can understand that by default you hide them, but having a way to reveal the prototypes is very convenient. As not all users can wait until the used software project changes the software to not use deprecated APIs.

As a second _user case_, we the tools developers nevertheless want to wrap as many API as possible. If we know, that if we set some magic variable and get all prototypes for all available symbols. We certainly will use it.

If `OMPI_OMIT_MPI1_COMPAT_DECLS` is that hook, to get all prototypes, than its something for the _users_.

Reply:
@gpaulsen I think the point here is that `OMPI_BUILDING` is a flag we use internally -- it means that additional things should be available because we are building Open MPI itself (vs. a user MPI application).  I think @bertwesarg is saying that `OMPI_BUILDING` should probably *also* imply that the MPI1 compatibility prototypes should be available if `OMPI_BUILDING==1`.  That way, `OMPI_OMIT_MPI1_COMPAT_DECLS` is purely tied to `--enable-mpi-compatibility` (i.e., solely "for the users").

Reply:
I see, let me see if I can do that and repush, for your review.